packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol

```diff
@@ -104,47 +119,49 @@ contract L2ToL2CrossDomainMessenger is IL2ToL2CrossDomainMessenger, ISemver, Tra
         if (_target == Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) revert MessageTargetL2ToL2CrossDomainMessenger();
 
-        bytes memory data = abi.encodeCall(
-            L2ToL2CrossDomainMessenger.relayMessage,
-            (_destination, block.chainid, messageNonce(), msg.sender, _target, _message)
-        );
-        assembly {
-            log0(add(data, 0x20), mload(data))
-        }
+        emit SentMessage(_destination, _target, messageNonce(), msg.sender, _message);
+
         msgNonce++;
     }
 
-    /// @notice Relays a message that was sent by the other CrossDomainMessenger contract. Can only be executed via
-    ///         cross-chain call from the other messenger OR if the message was already received once and is currently
-    ///         being replayed.
-    /// @param _destination Chain ID of the destination chain.
-    /// @param _source      Chain ID of the source chain.
-    /// @param _nonce       Nonce of the message being relayed.
-    /// @param _sender      Address of the user who sent the message.
-    /// @param _target      Address that the message is targeted at.
-    /// @param _message     Message payload to call target with.
+    /// @notice Relays a message that was sent by the other L2ToL2CrossDomainMessenger contract. Can only be executed
+    ///         via cross chain call from the other messenger OR if the message was already received once and is
+    ///         currently being replayed.
+    /// @param _id          Identifier of the SentMessage event to be relayed
+    /// @param _sentMessage Message payload of the `SentMessage` event
     function relayMessage(
-        uint256 _destination,
-        uint256 _source,
-        uint256 _nonce,
-        address _sender,
-        address _target,
-        bytes memory _message
+        ICrossL2Inbox.Identifier calldata _id,
+        bytes calldata _sentMessage
     )
         external
         payable
         nonReentrant
     {
-        if (msg.sender != Predeploys.CROSS_L2_INBOX) revert RelayMessageCallerNotCrossL2Inbox();
-        if (CrossL2Inbox(Predeploys.CROSS_L2_INBOX).origin() != Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) {
-            revert CrossL2InboxOriginNotL2ToL2CrossDomainMessenger();
+        // Ensure the log came from the messenger. Since the log origin is the CDM, there isn't a scenario where a
+        // relayMessage is invoked from the CrossL2Inbox as the SentMessage log is not calldata for this function
+        if (_id.origin != Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) {
+            revert IdentifierOriginNotL2ToL2CrossDomainMessenger();
         }
+
+        // Signal that this is a cross chain call that needs to have the identifier validated
+        CrossL2Inbox(Predeploys.CROSS_L2_INBOX).validateMessage(_id, keccak256(_sentMessage));
+
+        // Decode the log topics
+        (bytes32 selector, uint256 _destination, address _target, uint256 _nonce) =
```

Comment:
either missing underscore on `selector` or extra underscores on the other 3 vars.
the current styleguide, which I don't really like but whatevs, indicates that variables should NOT be underscored, only arguments are underscored.
it'd be my pleasure to fight for underscored variables, since arguments are ALSO in-memory variables. (except calldata ones)

Reply:
Ahh okay, let me update this